content/blog/semgrep-post-message/index.md

```diff
@@ -0,0 +1,175 @@
+```
+
+We want to match function declarations (`pattern`) that **do not** contain any of our safe patterns (`pattern-not`). A few things are worth noting here:
+
+1. Our metavariable is matching on the context captured by `$FUNC`
+2. We need the the `function($OBJ) { ... }` check to exclude external function declarations (our second scenario)
+3. The `$OBJ` metavariable is matching the function parameter and making sure the `.origin` is being called on it
+
+We could likely use a `metavariable` for our equality operators instead of declaring all of these `pattern-not`, but I like how explicit the rule currently is. Time to match those external functions!
+
+#### External functions
+
+```yaml
+pattern-either:
+  - pattern: |
+      function $FNAME(...) { $CTX }
+      ...
+      window.addEventListener('message', $FNAME,...)
+  - pattern: |
+      $FNAME = (...) => { $CTX }
+      ...
+      window.addEventListener('message', $FNAME,...)
+```
+
+_This is the only situation where I couldn't use the same pattern to match `arrow functions` and normal `function` declarations. Let me know if I could simplify this pattern._
+
+Here we are making sure that we can match any of the function declaration types with `pattern-either` and capturing the function names during declaration with `$FNAME`. Note that we use the same `$FNAME` metavariable to ensure that the **same name** is being used within `addEventListener`. The ellipsis (`...`) are there to match anything in-between those two lines.
+
+We are also capturing the context of the function body itself with `$CTX` and we will use it to exclude our **safe patterns**. Let's get to it:
+
+```yaml
+pattern-either:
+  - pattern: |
+      function $FNAME(...) { $CTX }
+      ...
+      window.addEventListener('message', $FNAME,...)
+  - pattern: |
+      $FNAME = (...) => { $CTX }
+      ...
+      window.addEventListener('message', $FNAME,...)
+  - metavariable-pattern:
+    metavariable: $CTX
+    patterns:
+      - pattern-not: |
+          if ($OBJ.origin == $X) { ... }
+      - pattern-not: |
+          if ($OBJ.origin === $X) { ... }
+      - pattern-not: |
+          if ($OBJ.origin != $X) { ... }
+      - pattern-not: |
+          if ($OBJ.origin !== $X) { ... }
+      - pattern-not: |
+          if ($REGEX.test($OBJ.origin)) { ... }
+```
+<br />
+
+Nothing new here, we are ensuring that we ignore matches if they contain any of the `if` patterns declared by our `pattern-not` operators. These patterns are being checked on the context of `$CTX` which is our function body.
+
+With those two cases being taken care of separately we can move to our final rule declaration.
+
+## Final pattern
+
+In order to make this work we need to glue both scenario using `pattern-either`. Let's check our final rule in the playground:
+
+<iframe title="Rails Cache Invalid Args" src="https://semgrep.dev/embed/editor?snippet=bernardoamc:post_message_origin_check" width="100%" height="430px" frameborder="0"></iframe>
```

Comment:
This is super cool! Would you be interested in sharing your work with others via the Semgrep Registry by filing a [PR](https://github.com/returntocorp/semgrep-rules/pulls)?

Reply:
For sure, that was going to be my next question! :) I will adjust the rule based on your feedback and open the PR right after.

Reply:
https://github.com/returntocorp/semgrep-rules/pull/1415 :)